jina/kubernetes/kubernetes_tools.py

```diff
@@ -1,86 +1,130 @@
 import threading
 
 import pathlib
+from typing import Union
+
 from bcolors import UNDERLINE, ENDC, BOLD, PASS
 import time
 
+# from kubernetes import utils
 
-from kubernetes import client, config, utils
 import os
 import tempfile
 
-from kubernetes.utils import FailToCreateError
-from kubernetes.watch import Watch
+# from kubernetes.utils import FailToCreateError
+# from kubernetes.watch import Watch
 
 cur_dir = os.path.dirname(__file__)
 
 
-config.load_kube_config()
-k8s_client = client.ApiClient()
-v1 = client.CoreV1Api()
-beta = client.ExtensionsV1beta1Api()
-networking_v1_beta1_api = client.NetworkingV1beta1Api()
-
-
-def create_gateway_ingress(namespace: str):
-    # create ingress class
-    # body = client.V1beta1IngressClass(
-    #     api_version="networking.k8s.io/v1",
-    #     kind="IngressClass",
-    #     metadata=client.V1ObjectMeta(
-    #         name=f'{namespace}-lb',
-    #         annotations= {
-    #             "linkerd.io/inject": "enabled"
-    #         }
-    #     ),
-    #     spec=client.V1beta1IngressClassSpec(
-    #         controller="example.com/ingress-controller",
-    #         parameters={
-    #             "apiGroup": "k8s.example.com",
-    #             "kind": "IngressParameters",
-    #             "name": "external-lb",
-    #         }
-    #     )
-    # )
-    # networking_v1_beta1_api.create_ingress_class(
-    #     body=body
-    # )
-
-    # create ingress
-    body = client.NetworkingV1beta1Ingress(
-        api_version="networking.k8s.io/v1beta1",
-        kind="Ingress",
-        metadata=client.V1ObjectMeta(
-            name=f'{namespace}-ingress',
-            annotations={
-                "nginx.ingress.kubernetes.io/rewrite-target": "/",
-                # "linkerd.io/inject": "enabled",
-            },
-        ),
-        spec=client.NetworkingV1beta1IngressSpec(
-            rules=[
-                client.NetworkingV1beta1IngressRule(
-                    host="",
-                    http=client.NetworkingV1beta1HTTPIngressRuleValue(
-                        paths=[
-                            client.NetworkingV1beta1HTTPIngressPath(
-                                path="/",
-                                backend=client.NetworkingV1beta1IngressBackend(
-                                    service_port=8080, service_name="gateway-exposed"
-                                ),
-                            )
-                        ]
-                    ),
-                )
-            ]
-        ),
-    )
-    # Creation of the Deployment in specified namespace
-    # (Can replace "default" with a namespace you may have created)
-    networking_v1_beta1_api.create_namespaced_ingress(namespace=namespace, body=body)
+class ClientsSingelton:
```

Comment:
typo: singleton

Reply:
How do u make sure this is a `singleton`? I think the only way in Python is to implement a `metaclass` that controls the actual instantiation